### PR TITLE
Implement model ensembling features

### PR DESCRIPTION
**Pull request by RAMitchell**
Implements #167 by adding multiplcation and addition operators for our estimators.

Additional is carefully defined for every member variable in the estimator.

### CODE REVIEW

**Comment by seberg** [comment]
Looking through it a bit, it seems right and cute to me (but I should understand the actual prediction mixing a bit better).

But I also have to wonder a bit if this is actually all that useful compared to `VotingClassifier` and `VotingEstimator` in `sklearn`?  It seems a bit like a really cool idea that may not be quite as useful in practice.

OTOH, I am not really worried about it even if I am not sure it is the best pattern:  I don't think `+` or `*` on estimators can be used for anything else useful.

**Comment by RAMitchell** [comment]
Thanks for your quick review!

The voting classifier gives a different result. The method described here combines the result in logit space before probability transformation. It is not available in sklearn.

I have personally used these features before by hacking into xgboost e.g. for this paper:
https://arxiv.org/abs/2005.07353

**Comment by RAMitchell** [comment]
Its also another method of out-of-core batch training different from the example in https://github.com/rapidsai/legate-boost/tree/main/examples/batch_training. I think it has potential in memory constrained situations.

Its definitely up for debate if my +/* operator overloading API is sensible.

**Comment by seberg** [comment]
It isn't the same as `voting="soft"`, even if I might agree that having a default of `voting="hard"` may not be the best choice (maybe should just think about it more)?

In either case, I think my main worry is that the `+`/`*` syntax may run into oddities.  Maybe just things like `model_a * 0.3 + model_b * 0.8` maybe not being flagged (factors don't sum to 1)?

File: legateboost/models/tree.py [hunk 1]
```diff
@@ -249,3 +240,8 @@ def recurse_print(id: int, depth: int) -> str:
             return text
 
         return recurse_print(0, 0)
+
+    def __mul__(self, scalar: Any) -> "Tree":
+        new = copy.deepcopy(self)
+        new.leaf_value *= scalar
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply legateboost/models/tree.py:246]: In a way, assuming the dataset is the same or following the same distribution, is this the same as changing the learning rate during training?

File: legateboost/legateboost.py [hunk 2]
```diff
@@ -54,6 +54,29 @@ def __init__(
...
         self.base_models = base_models
 
+        # define what happens to the attributes when two models are added
+        self._add_behaviour.update(
+            {
+                "models_": AddMember.ADD,
+                "model_init_": AddMember.ADD,
+                "n_features_in_": AddMember.ASSERT_SAME,
+                "is_fitted_": AddMember.ASSERT_SAME,
+                "n_estimators": AddMember.PREFER_A,
```
Comment (trivialfis) [inline legateboost/legateboost.py:64]: shouldn't this be `ADD`?

File: legateboost/models/tree.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RAMitchell) [reply legateboost/models/tree.py:246]: Yes it makes sense for all objectives. The leaf output is always additive until some (optional) non-linear transformation gets added at the end of prediction.

The multiplication by a scalar is similar to learning rate, in the case of training two models independently with learning rate 0.5 and adding them, they will not be the same as one model trained with learning rate 1.0. At each step the gradients are calculated without access to the other models predictions.

File: legateboost/legateboost.py (hunk 2 above)
Comment (RAMitchell) [reply legateboost/legateboost.py:64]: Yes I think it makes sense to be ADD here.

**Comment by RAMitchell** [comment]
@seberg I don't think the scalar being greater than 1 is a problem - it would be a legitimate thing to do, although I don't know why you would do it. Gradient boosted models are linearly additive and this is just expressing that.

**Review by trivialfis — APPROVED** [review]
Looks good to me, a very interesting feature!

Some notes:
- It might not be necessary to have the same model types. Some people stack models.
- The addition is not associative, which is weird from an algebra standpoint, but it's fine as an interface for ML models.

**Comment by RAMitchell** [comment]
@trivialfis In which way is it not associative?

**Comment by trivialfis** [comment]
Sorry, commutative

**Comment by RAMitchell** [comment]
Isn't it also commutative. Do you have an example where its not?

**Comment by trivialfis** [comment]
If you have prefer A option, then the ordering of the addition matters right?

**Comment by RAMitchell** [comment]
Ah yes for the parameters. The models raw prediction output is associative and commutiative though.

**Comment by trivialfis** [comment]
Yeah, should be fine as an interface. If one really wants to be rigorous, he can simply discard those parameters. Just something to be aware of, not important. The PR looks great, very interesting feature!
